packages/@expo/config-plugins/src/utils/plugin-resolver.ts

```diff
@@ -10,26 +9,41 @@ import { ConfigPlugin, StaticPlugin } from '../Plugin.types';
 // Default plugin entry file name.
 export const pluginFileName = 'app.plugin.js';
 
-function findUpPackageJson(root: string): string {
-  const packageJson = findUp.sync('package.json', { cwd: root });
-  assert(packageJson, `No package.json found for module "${root}"`);
-  return packageJson;
-}
+// pluginReference is a node module or a file path, as user entered it in app.config.js
+export function resolvePluginForModule(
+  projectRoot: string,
+  pluginReference: string
+): { filePath: string; isPluginFile: boolean } {
+  const resolvedPackageJson = resolveFrom.silent(projectRoot, `${pluginReference}/package.json`);
+  const resolvedPluginScriptFile = resolveFrom.silent(projectRoot, pluginReference);
+  const hasResolved = !!(resolvedPackageJson || resolvedPluginScriptFile);
 
-export function resolvePluginForModule(projectRoot: string, modulePath: string) {
-  const resolved = resolveFrom.silent(projectRoot, modulePath);
-  if (!resolved) {
+  function throwPluginNotFound(): never {
     throw new PluginError(
-      `Failed to resolve plugin for module "${modulePath}" relative to "${projectRoot}"`,
+      `Failed to resolve plugin for module "${pluginReference}" relative to "${projectRoot}"`,
       'PLUGIN_NOT_FOUND'
     );
   }
-  // If the modulePath is something like `@bacon/package/index.js` or `expo-foo/build/app`
+  if (!hasResolved) {
+    throwPluginNotFound();
+  }
+  // If the pluginReference is something like `@bacon/package/index.js` or `expo-foo/build/app`
   // then skip resolving the module `app.plugin.js`
-  if (moduleNameIsDirectFileReference(modulePath)) {
-    return { isPluginFile: false, filePath: resolved };
+  if (moduleNameIsDirectFileReference(pluginReference) && resolvedPluginScriptFile) {
+    return { isPluginFile: false, filePath: resolvedPluginScriptFile };
+  }
+  if (!resolvedPackageJson) {
+    throwPluginNotFound();
+  }
+  const maybePluginFilePath = findUpPlugin(resolvedPackageJson);
+  if (maybePluginFilePath) {
+    return { isPluginFile: true, filePath: maybePluginFilePath };
+  } else {
+    if (!resolvedPluginScriptFile) {
+      throwPluginNotFound();
+    }
+    return { isPluginFile: false, filePath: resolvedPluginScriptFile };
```

Comment:
I'd love to refactor this a bit. The way I like to think of "returning functions", where some data in should return some other data out, is that we can structure it like this:

```
function <name>(a, b) {
  // Handle most common case
  // Handle alternative cases
  // return fallback, or throw
}
```

It would make the function way easier to read and optimize, for example:

```ts
// pluginReference is a node module or a file path, as user entered it in app.config.js
export function resolvePluginForModule(
  projectRoot: string,
  pluginReference: string
): { filePath: string; isPluginFile: boolean } {
  // If the plugin was referenced by direct file reference, or package + file specifier, skip resolving app.plugin.js
  if (moduleNameIsDirectFileReference(pluginReference)) {
    const pluginScriptFile = resolveFrom.silent(projectRoot, pluginReference);
    if (pluginScriptFile) {
      return { isPluginFile: false, filePath: pluginScriptFile };
    }
  }

  // If the plugin was referenced by node module name, resolve the root of the package and look for app.plugin.js
  const pluginPackageDir = resolveFrom.silent(projectRoot, `${pluginReference}/package.json`);
  const pluginPackageFile = pluginPackageDir ? findUpPlugin(pluginPackageDir) : null;
  if (pluginPackageFile) {
    return { isPluginFile: true, filePath: pluginPackageFile };
  }

  throw new PluginError(
    `Failed to resolve plugin for module "${pluginReference}" relative to "${projectRoot}"`,
    'PLUGIN_NOT_FOUND'
  );
}
```

Reply:
It also shows further optimization we could do:

```ts
  // If the plugin was referenced by node module name, resolve the root of the package and look for app.plugin.js
  const pluginPackageFile = resolveFrom.silent(projectRoot, `${pluginReference}/${pluginFileName}`);
  if (pluginPackageFile && fileExists(pluginPackageFile)) {
    return { isPluginFile: true, filePath: pluginPackageFile };
  }
```

Instead of resolving this order:
  - `root = resolve(<pkg>/package.json)`
  - `plugin = resolve(path.dirname(root), './app.plugin.js')`
  - `return fileExists(plugin)`

Reply:
Oh, another optimization to really get the most out of it would be:

```
function resolvePluginForModule(
  projectRoot: string,
  pluginReference: string
): { filePath: string; isPluginFile: boolean } {
  if (moduleNameIsDirectFileReference(pluginReference)) {
    // Only resolve `./file.js`, `package/file.js`, `@org/package/file.js`
  } else if (moduleNameIsPackageReference(pluginReference)) {
    // Only resolve `package -> package/app.plugin.js`, `@org/package -> 
  @org/package/app.plugin.js` 
  }

  throw new PluginError(
    `Failed to resolve plugin for module "${pluginReference}" relative to "${projectRoot}"`,
    'PLUGIN_NOT_FOUND'
  );
}
```